include/onnx-mlir/Runtime/OMTensor.h

```diff
@@ -0,0 +1,387 @@
+ *         NULL if the data buffer is not set.
+ */
+void *omTensorGetAlignedPtr(OMTensor *tensor);
+
+/**
+ * OMTensor allocated and aligned pointer setter.
+ *
+ * @param tensor pointer to the OMTensor
+ * @param owning whether allocatedPtr should be freed after tensor is destroyed.
+ * @param allocatedPtr allocated pointer to tensor content.
+ * @param alignedPtr aligned pointer to tensor content. If NULL will be set to
+ * allocatedPtr.
+ *
+ *
+ */
+void omTensorSetPtr(
+    OMTensor *tensor, int owning, void *allocatedPtr, void *alignedPtr);
+
+/**
+ * OMTensor allocated ptr getter.
+ *
+ * @param tensor pointer to the OMTensor
+ * @return pointer to the allocated data buffer of the OMTensor,
+ *         NULL if the allocated data buffer is not set.
+ */
+void *omTensorGetAllocatedPtr(OMTensor *tensor);
+
+/**
+ * OMTensor data sizes getter
+ *
+ * @param tensor pointer to the OMTensor
+ * @return pointer to the data shape array.
+ */
+int64_t *omTensorGetDataShape(OMTensor *tensor);
+
+/**
+ * OMTensor data sizes setter
+ *
+ * @param tensor pointer to the OMTensor
+ * @param shape data sizes array to be set
+ *
+ * Set the data sizes array of the OMTensor to the values in the input array.
+ */
+void omTensorSetShape(OMTensor *tensor, int64_t *shape);
+
+/**
+ * OMTensor data strides getter
+ *
+ * @param tensor pointer to the OMTensor
+ * @return pointer to the data strides array.
+ */
+int64_t *omTensorGetStrides(OMTensor *tensor);
+
+/**
+ * OMTensor data strides setter
+ *
+ * @param tensor pointer to the OMTensor
+ * @param strides tensor strides array to be set.
+ *
+ * Set the data strides array of the OMTensor to the values in the input array.
+ */
+void omTensorSetStrides(OMTensor *tensor, int64_t *strides);
+
+/**
+ * OMTensor data type getter
+ *
+ * @param tensor pointer to the OMTensor
+ * @return ONNX data type of the data buffer elements.
+ */
+OM_DATA_TYPE omTensorGetDataType(OMTensor *tensor);
+
+/**
+ * OMTensor data type setter
+ *
+ * @param tensor pointer to the OMTensor
+ * @param dataType ONNX data type to be set
+ *
+ * Set the ONNX data type of the data buffer elements.
+ */
+void omTensorSetDataType(OMTensor *tensor, OM_DATA_TYPE dataType);
+
+/* Helper function to get the ONNX data type size in bytes */
+static inline int getDataTypeSize(OM_DATA_TYPE dataType) {
+  return OM_DATA_TYPE_SIZE[dataType];
+}
+
+/**
+ * OMTensor data buffer size getter
+ *
+ * @param tensor pointer to the OMTensor
+ * @return the total size of the data buffer in bytes.
+ */
+int64_t omTensorGetDataBufferSize(OMTensor *tensor);
+
+/**
+ * OMTensor rank getter
+ *
+ * @param tensor, pointer to the OMTensor
+ * @return rank of data sizes and strides of the OMTensor.
+ */
+int omTensorGetRank(OMTensor *tensor);
+
+/**
+ * OMTensor number of elements getter
+ *
+ * @param tensor, pointer to the OMTensor
+ * @return the number of elements in the data buffer.
+ */
+int64_t omTensorGetNumElems(OMTensor *tensor);
+
+#ifdef __cplusplus
```

Comment:
Also, the `#ifdef __cplusplus` surrounded code are meant to be "internal" only, i.e., they are only called by ONNX MLIR itself, not meant for end users. So I don't think they should appear in user header OMTensor.h.

Reply:
Currently, there're only a subset of C++ APIs used by the numerical tests. As per our (Gong and I) discussion, since they provide value for potential use case, they can stay inside the OMTensor header. 

Let me try to clean things up a bit by moving these unpolished C++ functions to a separate file inside an internal directory for the moment. In the future, I will:
- Provide full API spec for C++ user operating on OMTensor.
- Document them to the same level of detailedness as C99 API.
- Try to move implementation out of the header (this may be difficult with template going on).